twine: add page

tjsmart opened the pull request:
<!--
Thank you for contributing!
Please fill in the following checklist, removing items that do not apply.
See also https://github.com/tldr-pages/tldr/blob/main/CONTRIBUTING.md
-->

Closes #8430

- [x] The page(s) are in the correct platform directories: `common`, `linux`, `osx`, `windows`, `sunos`, `android`, etc.
- [x] The page(s) have at most 8 examples.
- [x] The page description(s) have links to documentation or a homepage.
- [x] The page(s) follow the [content guidelines](/tldr-pages/tldr/blob/main/CONTRIBUTING.md#guidelines).
- [x] The PR title conforms to the recommended [templates](/tldr-pages/tldr/blob/main/CONTRIBUTING.md#commit-message).
- **Version of the command being documented (if known):**

```
$ twine --version
twine version 4.0.1 (importlib-metadata: 4.13.0, keyring: 23.9.3, pkginfo: 1.8.3, requests: 2.28.1, requests-toolbelt: 0.9.1, urllib3:
1.26.12)
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/tldr-pages/tldr?pullRequest=8624) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/tldr-pages/tldr?pullRequest=8624) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/tldr-pages/tldr?pullRequest=8624) it.</sub>

kbdharun reviewed (commented):
Welcome to tldr @tjsmart. I have some suggestions for this page.

> Tip: Try to use mnemonics whenever possible

tjsmart commented on pages/common/twine.md:
```diff
@@ -0,0 +1,21 @@
+> Subcommands such as `twine upload` have their own usage documentation.
+> More information: <https://twine.readthedocs.io>.
+
+- Upload to Test PyPI and verify things look right:
```
The 'r' stands for 'repository'. So in the edit I changed this line to be:
```
- Upload to the Test PyPI [r]epository and verify things look right:
```

tjsmart commented on pages/common/twine.md:
```diff
@@ -0,0 +1,21 @@
+
+`twine upload dist/*`
+
+- Specify username and password:
```
Applied this locally, thanks!

tjsmart commented:
> Welcome to tldr @tjsmart. I have some suggestions for this page.
> 
> > Tip: Try to use mnemonics whenever possible

Thanks @kbdharun for the welcome and the suggestions!

kbdharun commented on pages/common/twine.md:
```diff
@@ -0,0 +1,21 @@
+# twine
+
+> Twine is a utility for publishing Python packages on PyPI.
+> Subcommands such as `twine upload` have their own usage documentation.
```
What is the difference between the current command and `twine-upload`? Is this line necessary?

tjsmart replied:
I used `pages/common/pip.md` as a template which had a line similar to this one.

`twine` has two other subcommands `check` and `register`, which I believe are seldom used. But perhaps I should include one example of each?

kbdharun replied:
> I used `pages/common/pip.md` as a template which had a line similar to this one.
> 
> `twine` has two other subcommands `check` and `register`, which I believe are seldom used. But perhaps I should include one example of each?

Thanks for the information, I will mark this conversation as resolved. As this line is applicable here.

corneliusroemer replied:
There does not seem to be a `tldr` page `twine upload` so it's weird to refer to it as if it existed.

I would get rid of this sentence here, and if applicable/useful, add one point for usage of `check` and `register` each.

If this was ever broken out, we could then add `Subcommands such as `twine upload` have their own usage documentation.` but right now this will just confuse users, as they won't find a `tldr twine upload` page.

navarroaxel commented on pages/common/twine.md:
```diff
@@ -0,0 +1,21 @@
+
+`twine upload dist/*`
+
+- Specify [u]sername and [p]assword while uploading to PyPI:
```
```suggestion
- Upload to PyPI with a specified [u]sername and [p]assword:
```

navarroaxel commented on pages/common/twine.md:
```diff
@@ -0,0 +1,21 @@
+# twine
+
+> Twine is a utility for publishing Python packages on PyPI.
```
Just describe the command
```suggestion
> Utility for publishing Python packages on PyPI.
```

navarroaxel commented on pages/common/twine.md:
```diff
@@ -0,0 +1,21 @@
+
+`twine upload -u {{username}} -p {{password}} dist/*`
+
+- Upload to an alternative repository url:
```
alternative like a mirror? or another registry?
```suggestion
- Upload to an alternative repository URL:
```

tjsmart replied:
Could be any valid repository (package index) URL. For some cases, you might only upload to a single url:
```
twine upload --repository-url <my pypi url> dist/*
```

Or perhaps issue twine upload multiple times to upload to different urls
```
twine upload --repository-url <my first pypi url> dist/*
twine upload --repository-url <my second pypi url> dist/*
```

corneliusroemer commented on pages/common/twine.md:
```diff
@@ -0,0 +1,21 @@
+
+`twine upload -r testpypi dist/*`
+
+- Upload to PyPI:
```
I would put the Upload to PyPI first, before test upload.

The upload itself is after all the main usage, the test upload is optional.

corneliusroemer reviewed (commented):
For reference for others reviewing this, here's the full help `twine upload -h`

Maybe the following would be useful as well:
`--config-file CONFIG_FILE The .pypirc config file to use.`
`--skip-existing       Continue uploading files if one already exists. (Only valid when uploading to PyPI. Other implementations may not
                        support this.)`: may work similar to a force?
`--verbose  `

```
twine upload -h                 
usage: twine upload [-h] [-r REPOSITORY] [--repository-url REPOSITORY_URL] [-s] [--sign-with SIGN_WITH] [-i IDENTITY] [-u USERNAME]
                    [-p PASSWORD] [--non-interactive] [-c COMMENT] [--config-file CONFIG_FILE] [--skip-existing] [--cert path]
                    [--client-cert path] [--verbose] [--disable-progress-bar]
                    dist [dist ...]

positional arguments:
  dist                  The distribution files to upload to the repository (package index). Usually dist/* . May additionally contain a
                        .asc file to include an existing signature with the file upload.

options:
  -h, --help            show this help message and exit
  -r REPOSITORY, --repository REPOSITORY
                        The repository (package index) to upload the package to. Should be a section in the config file (default: pypi).
                        (Can also be set via TWINE_REPOSITORY environment variable.)
  --repository-url REPOSITORY_URL
                        The repository (package index) URL to upload the package to. This overrides --repository. (Can also be set via
                        TWINE_REPOSITORY_URL environment variable.)
  -s, --sign            Sign files to upload using GPG.
  --sign-with SIGN_WITH
                        GPG program used to sign uploads (default: gpg).
  -i IDENTITY, --identity IDENTITY
                        GPG identity used to sign files.
  -u USERNAME, --username USERNAME
                        The username to authenticate to the repository (package index) as. (Can also be set via TWINE_USERNAME environment
                        variable.)
  -p PASSWORD, --password PASSWORD
                        The password to authenticate to the repository (package index) with. (Can also be set via TWINE_PASSWORD
                        environment variable.)
  --non-interactive     Do not interactively prompt for username/password if the required credentials are missing.
  -c COMMENT, --comment COMMENT
                        The comment to include with the distribution file.
  --config-file CONFIG_FILE
                        The .pypirc config file to use.
  --skip-existing       Continue uploading files if one already exists. (Only valid when uploading to PyPI. Other implementations may not
                        support this.)
  --cert path           Path to alternate CA bundle (can also be set via TWINE_CERT environment variable).
  --client-cert path    Path to SSL client certificate, a single file containing the private key and the certificate in PEM format.
  --verbose             Show verbose output.
  --disable-progress-bar
                        Disable the progress bar.
```

For check:
```
usage: twine check [-h] dist [dist ...]

positional arguments:
  dist        The distribution files to check, usually dist/*
```

`register` seems not to be very mainstream, so we can leave it out:
```
register operation is not required with PyPI.org
```

corneliusroemer commented:
I can't unresolve this conversation apparently:
![image](https://user-images.githubusercontent.com/25161793/194855678-7dd19a8d-8a2b-4566-a3c4-7d656c197d50.png)

There does not seem to be a tldr page twine upload so it's weird to refer to it as if it existed. Or am I missing something.

I would get rid of this sentence here, and if applicable/useful, add one point for usage of check and register each.

If this was ever broken out, we could then add subcommands such as twine upload have their own usage documentation. but right now this will just confuse users, as they won't find a tldr twine upload page.

tjsmart commented:
Thanks for the suggestions @navarroaxel and @corneliusroemer !

@navarroaxel, I hadn't yet addressed the comments from @corneliusroemer...

I don't mind opening a separate PR to include @corneliusroemer suggestions since this one was merged.

Thoughts?

navarroaxel commented:
Yeah, open a new one @tjsmart

corneliusroemer commented:
@navarroaxel It's a bit confusing when there's still a conversation going on and it's merged already. At the same time I remember some of my proposals hanging there for a long time. So I can see benefits of both, fast merge if already useful and slow merge to resolve comments.

navarroaxel commented:
@corneliusroemer Yeah, I know but we are in hacktoberfest and it's an special moment of the year, most of the people sending PRs on this month aren't being seen the rest of the year. I expected more reviewers on this event. My fault.

corneliusroemer commented:
No worries at all! Wasn't aware of the hacktoberfest thing, that's a great justification for fast merging!
